share/models/change.py

```diff
@@ -271,3 +242,66 @@ def _resolve_change(self):
             else:
                 change[k] = v
         return change
+
+    def _merge_objects(self, from_obj, into_obj):
+        from share.models import ShareObject
+
+        # TODO less hacky way of expressing what can be merged
+        MERGEABLE_MODELS = {'AbstractCreativeWork', 'AbstractAgent', 'Award', 'AbstractWorkRelation', 'AbstractAgentRelation', 'AbstractAgentWorkRelation', 'ThroughTags', 'ThroughSubjects', 'ThroughContributor', 'ThroughAwards'}
+        concrete_model = from_obj._meta.concrete_model
+        if concrete_model.__name__ not in MERGEABLE_MODELS:
+            raise InvalidMergeError('Invalid model for merging: %s' % concrete_model)
+        if into_obj._meta.concrete_model is not concrete_model:
+            raise InvalidMergeError('Cannot merge objects of different concrete types: %s and %s' % (from_obj, into_obj))
+
+        # Avoid same_as chains
+        if into_obj.same_as_id:
+            into_obj = concrete_model.objects.get_canonical(into_obj.id)
+        concrete_model.objects.filter(same_as_id=from_obj.id).update(
+            change=self,
+            same_as=into_obj,
+            same_as_version=into_obj.version
+        )
+
+        for field in concrete_model._meta.get_fields():
+            if field.is_relation and not field.many_to_many and issubclass(field.remote_field.model, ShareObject) and hasattr(field, 'field'):
+                # Update incoming foreign keys to point to into_obj.
+                self._merge_fk_field(from_obj, into_obj, field.field)
+            elif not field.is_relation and field.editable and not field.primary_key:
+                # Update scalar field on into_obj. Use value from more recently modified object, ignoring blank values.
+                self._merge_scalar_field(from_obj, into_obj, field)
+
+        into_obj.change = self
+        into_obj.save()
+
+        from_obj.change = self
+        from_obj.same_as = into_obj
+        from_obj.same_as_version = into_obj.version
+        from_obj.save()
+
+        return into_obj
+
+    def _merge_fk_field(self, from_obj, into_obj, field):
+        for obj in field.model.objects.filter(**{field.name: from_obj.id}):
```

Comment:
There's no case where we want an insert, though... if there's no conflict, it's just an update. If there is a conflict, there are two updates (one to merge field values into the winner of the conflict, one to update the other's `same_as`). I made it recurse into `_merge_objects` so relation fields would be merged by the same logic. And just in case, say, we're merging an Agent with duplicate/conflicting Funder relations with associated Awards...